File: src/main/kotlin/macgram/blackjackbackend/domain/Game.java
package macgram.blackjackbackend.domain;

import macgram.blackjackbackend.domain.card.Card;
import macgram.blackjackbackend.domain.card.Cards;
import macgram.blackjackbackend.domain.user.Dealer;
import macgram.blackjackbackend.domain.user.Gamer;
import macgram.blackjackbackend.domain.user.Player;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class Game {
    private static final int INIT = 0;
    private static final int MAX = 2;

    private final Dealer dealer = new Dealer();
    private final List<Player> players = new ArrayList<>();
    private Queue<Card> cards;

    public void addPlayer(final Player player) {
        players.add(player);
    }

    public void removePlayer(final Player player) {
        players.remove(player);
    }

    public void start() {
        validateStart();
        cards = generateCards();
        distributeCard();
    }

    public void validateStart() {
        if (players.stream().anyMatch(it -> it.getBattingMoney() <= 0)) {
            throw new IllegalArgumentException("배팅하지 않은 사용자가 존재 합니다.");
        }
    }

    public void continueDealerCard() {
        dealer.continueCard(cards);
    }

    public boolean isEnded() {
        return players.stream()
                .noneMatch(Player::isContinued);
    }

    private void distributeCard() {
        for (int i = INIT; i < MAX; i++) {
            addDealerCard();
            addPlayersCard();
        }
    }

    public void addDealerCard() {
        addCard(dealer);
    }

    public void addPlayersCard() {
        players.stream()
                .filter(Player::isContinued)
                .forEach(this::addCard);
    }

    private void addCard(final Gamer gamer) {
        gamer.addCard(cards.poll());
    }

    public void mola() {
        int calculate = dealer.calculate();
    }

    private Queue<Card> generateCards() {
        Queue<Card> cards = new LinkedList<>();
        cards.addAll(Cards.shuffle());

        return cards;
    }

    public Dealer getDealer() {
        return dealer;
    }

    public List<Player> getPlayers() {
        return players;
    }
}

File: src/main/kotlin/macgram/blackjackbackend/domain/Game.kt
package macgram.blackjackbackend.domain

import macgram.blackjackbackend.domain.user.Dealer
import macgram.blackjackbackend.domain.user.Gamer
import java.util.ArrayList

import macgram.blackjackbackend.domain.card.Deck
import macgram.blackjackbackend.domain.user.Player


class Game {
    private val INIT: Int = 0
    private val MAX: Int = 2

    val dealer: Dealer = Dealer()
    var players: MutableList<Player> = ArrayList()
    lateinit var deck: Deck

    fun addPlayer(player: Player) {
        players.add(player)
    }

    fun removePlayer(player: Player) {
        players.remove(player)
    }

    fun start() {
        validateStart()
        deck = generateCards()
        deck.shuffle()
        distributeCard()
        println(dealer)
        println(players)

    }

    fun validateStart() {
        if (players.any { it.battingMoney <= 0 }) {
            throw IllegalArgumentException("배팅하지 않은 사용자가 존재 합니다.")
        }
    }

    fun continueDealerCard() {
        dealer.continueCard(deck)
    }

    fun isEnded(): Boolean {
        return players.none { it.continuable }
    }

    fun distributeCard() {
        for (i in INIT until MAX) {
            addDealerCard()
            addPlayersCard()
        }
    }

    fun addDealerCard() {
        addCard(dealer)
    }

    fun addPlayersCard() {
        players.filter(Player::continuable)
            .forEach(this::addCard)
    }

    fun generateCards(): Deck {
        return Deck()
    }

    private fun addCard(gamer: Gamer) {
        gamer.addCard(deck.pop())
    }
}
